File: 648. Replace Words/Python/Replace Words.py
class Solution:
    def replaceWords(self, dict, sentence):
        """
        :type dict: List[str]
        :type sentence: str
        :rtype: str
        """

        class TrieNode:
            def __init__(self, val):
                self.val = val
                self.children = [None] * 26
                self.is_word = False

        def build_trie(words):
            root = TrieNode(' ')
            for word in words:
                current = root
                for letter in word:
                    index = ord(letter) - ord('a')
                    if not current.children[index]:
                        current.children[index] = TrieNode(letter)
                    current = current.children[index]
                current.is_word = True
            return root

        def get_root(word, trie_root):
            current = ''
            for letter in word:
                current += letter
                index = ord(letter) - ord('a')
                if not trie_root.children[index]:
                    return word
                if trie_root.children[index].is_word:
                    return current
                trie_root = trie_root.children[index]
            return word

        trie_root = build_trie(dict)
        result = ''
        for word in sentence.split():
            result += get_root(word, trie_root) + ' '
        return result[:-1]

File: 648. Replace Words/Java/Replace Words.java
class Solution {
    public String replaceWords(List<String> dict, String sentence) {
        class TrieNode {
            char val;
            TrieNode[] children;
            boolean isWord;

            public TrieNode(char val) {
                this.val = val;
                children = new TrieNode[26];
                isWord = false;
            }
        }
        class Utils {
            TrieNode buildTrie(List<String> words) {
                TrieNode root = new TrieNode(' ');
                for (String word : words) {
                    TrieNode current = root;
                    for (char letter : word.toCharArray()) {
                        int index = letter - 'a';
                        if (current.children[index] == null) {
                            current.children[index] = new TrieNode(letter);
                        }
                        current = current.children[index];
                    }
                    current.isWord = true;
                }
                return root;
            }

            String getRoot(String word, TrieNode trieRoot) {
                StringBuilder builder = new StringBuilder();
                for (char letter : word.toCharArray()) {
                    builder.append(letter);
                    int index = letter - 'a';
                    if (trieRoot.children[index] == null) {
                        return word;
                    }
                    if (trieRoot.children[index].isWord) {
                        return builder.toString();
                    }
                    trieRoot = trieRoot.children[index];
                }
                return word;
            }
        }
        Utils utils = new Utils();
        TrieNode trieRoot = utils.buildTrie(dict);
        StringBuilder builder = new StringBuilder();
        for (String word : sentence.split(" ")) {
            builder.append(utils.getRoot(word, trieRoot)).append(' ');
        }
        builder.deleteCharAt(builder.length() - 1);
        return builder.toString();
    }
}